subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java

```diff
@@ -0,0 +1,140 @@
+ */
+package org.gradle.api.internal.artifacts.dependencies;
+
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.FileCollectionDependency;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.file.FileCollectionInternal;
+import org.gradle.api.tasks.TaskDependency;
+import org.gradle.internal.deprecation.DeprecationLogger;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Set;
+
+public class DefaultFileCollectionDependency extends AbstractDependency implements SelfResolvingDependencyInternal, FileCollectionDependency {
+    private final ComponentIdentifier targetComponentId;
+    private final FileCollectionInternal source;
+
+    public DefaultFileCollectionDependency(FileCollectionInternal source) {
+        this.targetComponentId = null;
+        this.source = source;
+    }
+
+    public DefaultFileCollectionDependency(ComponentIdentifier targetComponentId, FileCollectionInternal source) {
+        this.targetComponentId = targetComponentId;
+        this.source = source;
+    }
+
+    @Override
+    public boolean contentEquals(Dependency dependency) {
+        if (!(dependency instanceof DefaultFileCollectionDependency)) {
+            return false;
+        }
+        DefaultFileCollectionDependency selfResolvingDependency = (DefaultFileCollectionDependency) dependency;
+        return source.equals(selfResolvingDependency.source);
+    }
+
+    @Override
+    public DefaultFileCollectionDependency copy() {
+        return new DefaultFileCollectionDependency(targetComponentId, source);
+    }
+
+    @Nullable
+    @Override
+    public ComponentIdentifier getTargetComponentId() {
+        return targetComponentId;
+    }
+
+    @Override
+    public String getGroup() {
+        return null;
+    }
+
+    @Override
+    public String getName() {
+        return "unspecified";
+    }
+
+    @Override
+    public String getVersion() {
+        return null;
+    }
+
+    @Override
+    @Deprecated
+    @SuppressWarnings("deprecation")
+    public void resolve(org.gradle.api.internal.artifacts.CachingDependencyResolveContext context) {
+        context.add(source);
+    }
+
+    @Override
+    @Deprecated
+    public Set<File> resolve() {
+        if (emitDeprecation()) {
+            DeprecationLogger.deprecate("Directly resolving a file collection dependency's files")
+                .withAdvice("Add the dependency to a resolvable configuration and resolve the configuration.")
+                .willBecomeAnErrorInGradle9()
+                .withUpgradeGuideSection(8, "deprecate_self_resolving_dependency")
+                .nagUser();
+        }
+
+        return source.getFiles();
+    }
+
+    @Override
+    @Deprecated
+    public Set<File> resolve(boolean transitive) {
+
+        DeprecationLogger.deprecate("Directly resolving a file collection dependency's files")
```

Comment:
❓ Does this emit always because it's not used by AGP?

Reply:
Yes